### ISSUE TITLE
Provide information on VT-x and VT-d on x86_64

### ISSUE THREAD

**Issue by FCLC** [util-linux/util-linux#1603]
Currently when using default `lscpu` with no args, system will check for the VT-x extension support for host enabled hardware acceleration for guests.

An extension of VT-x is VT-d which allows for hardware pass through of devices in specific IOMMU groups.

In Intel parlance this is refereed to as "`Intel(R) Virtualization Technology for Directed I/O`" or VT-d for short. 

The easiest way to check if this is working correctly as of now is (as super user) `dmesg | grep DMAR`  and check for the early entry message of `Intel(R) Virtualization Technology for Directed I/O`

AMD equivalents to VT-x and VT-d are AMD-v and AMD-vi respectively.

I believe you currently use 
`CPUID.1:ECX.VMX[bit 5] = 1` to check for VMX. this should continue working.

**Comment by karelzak** [comment]
For now it's pretty simple (for VT-x)
```C
              if (cxt->virt->cpuflag && !strcmp(cxt->virt->cpuflag, "svm"))
                        add_summary_s(tb, sec, _("Virtualization:"), "AMD-V");
                else if (cxt->virt->cpuflag && !strcmp(cxt->virt->cpuflag, "vmx"))
                        add_summary_s(tb, sec, _("Virtualization:"), "VT-x");
```
So, the code checks for "vmx" and  "svm" CPU flags as specified in /proc/cpuinfo

The question is how to check for VT-d, because "dmesg | grep" is a pretty poor way (due to dmesg --clear, etc.). The ideal solution is /proc or /sys.

https://stackoverflow.com/questions/51261999/check-if-vt-d-iommu-has-been-enabled-in-the-bios-uefi
someone suggested "/sys/kernel/iommu_groups/*/devices/*", but I guess it works only when some devices are already configured.

Just for the record, some URLs about this topic
https://wiki.xenproject.org/wiki/VTd_HowTo
https://en.wikipedia.org/wiki/X86_virtualization#I/O_MMU_virtualization_(AMD-Vi_and_Intel_VT-d)

There is also VT-c (aka "connectivity") :-)

**Comment by FCLC** [comment]
> The question is how to check for VT-d, because "dmesg | grep" is a pretty poor way (due to dmesg --clear, etc.). The ideal solution is /proc or /sys.

We can check some of the sub flags of the VMX string in /proc/cpuinfo. 

I'll have to double check the flags but a none working VT-d flag output from /proc is 

`vmx flags	: vnmi preemption_timer posted_intr invvpid ept_x_only ept_ad ept_1gb flexpriority apicv tsc_offset vtpr mtf vapic ept vpid unrestricted_guest vapic_reg vid ple shadow_vmcs pml ept_mode_based_exec tsc_scaling usr_wait_pause`

Would have to compare to a system that only has VT-x available, compare against it and then use those?  

> The question is how to check for VT-d, because "dmesg | grep" is a pretty poor way (due to dmesg --clear, etc.). 

Agreed, more so It's been the most consistent I've found recently for casual usage on boot in bash scripts. 

Outside of the above, another way might be to try the suggestion from stack overflow for an existing device, if none is found attempt to add one as a way to probe the capability? 

If it succeeds, remove it and mark VT-d as enabled and functioning?  

Otherwise mark as not available?

**Comment by mator** [comment]
> Outside of the above, another way might be to try the suggestion from stack overflow for an existing device, if none is found attempt to add one as a way to probe the capability?

Well, even though checking device pci flags for vt-d, it's not a proper way to declare CPU flag for vt-d, and we could probably look at intel developer docs for processor vt-d support, but i failed to find it at the first glance...

**Comment by FCLC** [comment]
> > Outside of the above, another way might be to try the suggestion from stack overflow for an existing device, if none is found attempt to add one as a way to probe the capability?
> 
> Well, even though checking device pci flags for vt-d, it's not a proper way to declare CPU flag for vt-d, and we could probably look at intel developer docs for processor vt-d support, but i failed to find it at the first glance...

Yeah, I glanced at Volume 2 and Volume 3 yesterday looking for it with no luck. 

I think I may have found it, but it isn't currently displayed in /proc. 

Would a PR be accepted with direct register checks in x86_64 asm? 

something along the lines of 

```
global GetVMXCapability
global GetVTDCapability

section .text

                        ; Checks CPUID for VMX capability
GetVMXCapability:
    push rbx        ; Store the callers RBX register value

    mov eax, 1      ; Place a 1 into EAX to load standard set of CPUID data when called
    cpuid           ; place CPUID info into EAX EBX ECX and EDX

    mov eax, ecx    ; move the ECX registers values into EAX
    shr eax, 5      ; Read bit 5 by shifting to the right by 5 bits
    and eax, 1      ; and'ing the eax register with 1

    pop rbx         ; restore the callers RBX register value
    ret             ; Go back to calling routine; Checks CPUID for VMX capability


GetVTDCapability: 
    push rbx 
[
instructions here once I figure out which registers to verify
]
    pop rbx
    ret 
```

**Comment by mator** [comment]
> The question is how to check for VT-d, because "dmesg | grep" is a pretty poor way (due to dmesg --clear, etc.). The ideal solution is /proc or /sys.

per https://wiki.archlinux.org/title/PCI_passthrough_via_OVMF#Ensuring_that_the_groups_are_valid
we could "steal" code from `lspci` or use generic group description in `/sys/kernel/iommu_groups/*` or check for devices per https://lore.kernel.org/lkml/20140612215954.2754.30263.stgit@bling.home/

Having a real hardware with VT-D enabled and working, would be helpful... I could try to test it at home (though I have an older Xeon CPU)...

PS: I wasn't able to fetch http://www.intel.com/content/dam/www/public/us/en/documents/product-specifications/vt-directed-io-spec.pdf
